FLIGHT_DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,442 @@
+ 1. Throw an exception that derives from `RetryException`. When flight execution catches
+ that exception, it will attempt a retry.
+ 1. Return a StepResult of `STEP_RESULT_FAILURE_RETRY`.
+
+It is our experience that having the step perform its own exception handling and return
+the explicit `STEP_RESULT_FAILURE_RETRY` is more readable. It also keeps service
+exceptions separate from Stairway exceptions.
+
+Remember that simply requesting a retry and actually retrying are two separate
+things. Requesting retry asks Stairway to check the RetryRule to see if there are any more
+retries left. If the rule says, "no more retries", then the exception returned in the
+StepResult is used as the result exception of the flight.
+
+### Retry After Restart <a name="retryafterrestart"></a>
+Retrying is only in-memory. RetryRule state is not retained across a failure/recovery.
+
+Imagine that Stairway is running a step of a flight that uses a retry rule. The rule will
+try 10 times and then give up. Stairway has retried 9 times and then the server fails.
+When the flight is recovered, the step will be restarted and the retry rule counter will
+start at 0 once again.
+
+## Flight Construction <a name="flightconstruction"></a>
+
+The Stairway caller never constructs a Flight object. When you call Stairway to run a flight,
+you provide the input parameters and the class name. Stairway does the actual
+construction. That is because Stairway must be able to re-construct the flight after a
+failure/recovery from the data stored in the Stairway database.
+
+### Constructors Should... <a name="constructorsshould"></a>
+
+Your flight constructor code should construct and configure the steps and retry rules for
+the flight. The constructor might include:
+ - adding different steps based on input parameters
+ - extracting input parameters and providing them as constructor parameters to the steps
+
+### Constructors Should NOT... <a name="constructorsshouldnot"></a>
+
+The constructor **should not**:
+ - do database lookups
+ - access external services
+
+There are two reasons for those constraints. First, flight construction is done in
+Stairway on the recovery path. If a flight cannot be reconstructed, Stairway will
+fail to initialize. That can leave your service in a failed state.  
+
+Second, errors in flight construction can cause the flight to be marked as a dismal
+failure. It is better to keep the construction simpler and do the work within steps where
+you have control over the error handling, retrying and recovery.
+
+### Spring Patterns <a name="springpatterns"></a>
+
+Stairway does not use Spring, but it can be used by services that use Spring.
+
+When you construct a Stairway object, you can provide it with an _application
+context_. It is an `Object` that Stairway passes into flight constructors so they can fill in
+information about their service. It is easily confused with the Spring
```

Comment:
it's not really information about the service but dependencies on other services, clients, etc

Reply:
No, it is information about the service. In particular, in Spring, the binding of autowired classes is not available to flights. This is the way we make it available.